plugins/ingest/src/main/java/org/elasticsearch/ingest/processor/grok/GrokProcessor.java

```diff
@@ -0,0 +1,109 @@
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.ingest.processor.grok;
+
+import org.elasticsearch.ingest.Data;
+import org.elasticsearch.ingest.processor.ConfigException;
+import org.elasticsearch.ingest.processor.Processor;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.nio.charset.StandardCharsets;
+import java.util.Map;
+import java.util.Properties;
+
+public final class GrokProcessor implements Processor {
+
+    public static final String TYPE = "grok";
+    public static final String PATTERNS_PATH = "/org/elasticsearch/ingest/processor/grok";
+    public static final String PATTERN_PROPERTIES = "pattern.properties";
+
+    private final String matchField;
+    private final String matchPattern;
+    private Grok grok;
+
+    public GrokProcessor(String matchField, String matchPattern) throws ConfigException {
+        this.matchField = matchField;
+        this.matchPattern = matchPattern;
+
+        this.grok = new Grok();
+
+        // TODO(talevy): finalize a nicer way to manage these pattern files (properties file or filepaths)
+        try {
+            Properties patternProps = new Properties();
+            InputStream is = getClass().getResourceAsStream(PATTERNS_PATH + "/" + PATTERN_PROPERTIES);
+            patternProps.load(is);
+            for (String prop : patternProps.stringPropertyNames()) {
+                String patternFilename = patternProps.getProperty(prop);
+                this.grok.loadFromStream(getClass().getResourceAsStream(PATTERNS_PATH + "/patterns/" + patternFilename));
+            }
+        } catch (IOException e) {
+            throw new ConfigException(e);
+        }
+        this.grok.compile(this.matchPattern);
+    }
+
+    @Override
+    public void execute(Data data) {
+        Object field = data.getProperty(matchField);
+        // TODO(talevy): handle invalid field types
+        if (field instanceof String) {
+            Map<String, Object> matches = grok.captures((String) field);
+            if (matches != null) {
+                matches.forEach((k, v) -> data.addField(k, v));
+            }
+        }
+    }
+
+    public static class Builder implements Processor.Builder {
```

Comment:
yes, the builders are there to build a processor instance from a pipeline definition (which is converted to a map of maps). Additionally all the builders have setters to build a processor instance programmatically. Right now that is used only in tests but in the future other integrators can use that too.

Personally I think we can remove the builder and builder factory interfaces and replace it with a single factory interface (or another name? provider?) that knows how to build a processor instance from a map of maps. If we then also make the constructor of the processors public we can still build the processor instances in tests.